File: backend/app/api/v1/tenants.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from uuid import UUID

from app.db.session import get_db
from app.schemas.tenant import TenantCreate, TenantOut
from app.schemas.common import Page
from app.models.tenant import Tenant, TenantSettings
from app.models.billing import TenantSubscription, Plan
from app.core.config import settings
from app.api.deps import require_platform_admin, get_request_meta
from app.core.audit import make_audit_event
from app.services.template_packs import apply_pack_to_tenant
from app.services.finance_service import get_or_create_billing_profile, ensure_onboarding_row

router = APIRouter(prefix="/tenants")


@router.get("", response_model=Page[TenantOut])
def list_tenants(
    q: str | None = Query(default=None),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_db),
    admin = Depends(require_platform_admin),
):
    base = db.query(Tenant)
    if q:
        like = f"%{q.strip()}%"
        base = base.filter((Tenant.name.ilike(like)) | (Tenant.slug.ilike(like)))
    total = base.count()
    rows = (
        base.order_by(Tenant.created_at.desc())
        .offset(offset)
        .limit(limit)
        .all()
    )
    def _to_out(t: Tenant) -> TenantOut:
        return TenantOut(
            id=t.id,
            name=t.name,
            slug=t.slug,
            is_active=t.is_active,
            plan_key=(t.subscription.plan.key if (t.subscription and t.subscription.plan) else None),
            plan_name=(t.subscription.plan.name if (t.subscription and t.subscription.plan) else None),
            subscription_status=(t.subscription.status if t.subscription else None),
        )
    return Page(items=[_to_out(t) for t in rows], total=total, limit=limit, offset=offset)

@router.post("", response_model=TenantOut)
def create_tenant(
    payload: TenantCreate,
    db: Session = Depends(get_db),
    admin = Depends(require_platform_admin),
    meta: dict = Depends(get_request_meta),
):
    tenant = Tenant(name=payload.name, slug=payload.slug, is_active=True)
    db.add(tenant)
    db.flush()

    ts = TenantSettings(tenant_id=tenant.id, min_anon_threshold=settings.DEFAULT_MIN_ANON_THRESHOLD)
    db.add(ts)

    # Trial: aplica um plano default (START) para que features/limits funcionem imediatamente (DEV/onboarding).
    plan = db.query(Plan).filter(Plan.key == "START").first()
    if plan:
        sub = TenantSubscription(
            tenant_id=tenant.id,
            status="trial",
            plan_id=plan.id,
            entitlements_snapshot={"features": plan.features or {}, "limits": plan.limits or {}},
        )
    else:
        sub = TenantSubscription(
            tenant_id=tenant.id,
            status="trial",
            entitlements_snapshot={"features": {"LMS": True}, "limits": {}},
        )
    db.add(sub)

    billing_profile = get_or_create_billing_profile(db, tenant.id)
    billing_profile.legal_name = tenant.name
    billing_profile.trade_name = tenant.name
    db.add(billing_profile)

    ensure_onboarding_row(db, tenant.id)

    # Onboarding: aplica automaticamente o pack padrão de templates (NR-1) ao novo tenant.
    pack_key = (getattr(settings, 'AUTO_APPLY_TEMPLATE_PACK_KEY', '') or '').strip()
    if pack_key:
        apply_pack_to_tenant(db, pack_key=pack_key, tenant_id=tenant.id)

    ae = make_audit_event(
        tenant_id=None,
        actor_user_id=admin.id,
        action="CREATE",
        entity_type="TENANT",
        entity_id=tenant.id,
        before=None,
        after={"name": tenant.name},
        ip=meta.get("ip"),
        user_agent=meta.get("user_agent"),
        request_id=meta.get("request_id"),
    )
    db.add(ae)
    db.commit()
    db.refresh(tenant)
    return TenantOut(id=tenant.id, name=tenant.name, slug=tenant.slug, is_active=tenant.is_active, plan_key=(tenant.subscription.plan.key if (tenant.subscription and tenant.subscription.plan) else None), plan_name=(tenant.subscription.plan.name if (tenant.subscription and tenant.subscription.plan) else None), subscription_status=(tenant.subscription.status if tenant.subscription else None))


@router.get("/{tenant_id}", response_model=TenantOut)
def get_tenant(tenant_id: UUID, db: Session = Depends(get_db), admin = Depends(require_platform_admin)):
    t = db.query(Tenant).filter(Tenant.id == tenant_id).first()
    if not t:
        from app.core.errors import NotFound
        raise NotFound("Tenant não encontrado")
    return TenantOut(id=t.id, name=t.name, slug=t.slug, is_active=t.is_active, plan_key=(t.subscription.plan.key if (t.subscription and t.subscription.plan) else None), plan_name=(t.subscription.plan.name if (t.subscription and t.subscription.plan) else None), subscription_status=(t.subscription.status if t.subscription else None))

File: frontend/src/lib/api/tenants.ts
import { apiFetch } from "@/lib/api/client";
import type { Page } from "@/lib/api/types";

export type TenantOut = {
  id: string;
  name: string;
  slug?: string | null;
  is_active: boolean;
  plan_key?: string | null;
  plan_name?: string | null;
  subscription_status?: string | null;
};

export type TenantCreate = {
  name: string;
  slug?: string | null;
};

export type PlanAdminOut = {
  id: string;
  key: string;
  name: string;
  features: Record<string, any>;
  limits: Record<string, any>;
  price_monthly?: number | null;
  price_annual?: number | null;
  is_custom_price?: boolean;
  stripe_price_id?: string | null;
  stripe_price_id_monthly?: string | null;
  stripe_price_id_annual?: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export async function listTenants(q?: string, limit: number = 50, offset: number = 0) {
  const qp = new URLSearchParams();
  if (q) qp.set("q", q);
  qp.set("limit", String(limit));
  qp.set("offset", String(offset));
  return apiFetch<Page<TenantOut>>("console", `/tenants?${qp.toString()}`);
}

export async function createTenant(payload: TenantCreate) {
  return apiFetch<TenantOut>("console", "/tenants", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function listPlatformPlans() {
  return apiFetch<PlanAdminOut[]>("console", "/platform/plans");
}

export async function createPlan(payload: {
  key: string;
  name: string;
  features?: Record<string, any>;
  limits?: Record<string, any>;
  price_monthly?: number | null;
  price_annual?: number | null;
  is_custom_price?: boolean;
  stripe_price_id?: string | null;
  stripe_price_id_monthly?: string | null;
  stripe_price_id_annual?: string | null;
  is_active?: boolean;
}) {
  return apiFetch<PlanAdminOut>("console", "/platform/plans", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function updatePlan(planId: string, payload: {
  name?: string;
  features?: Record<string, any>;
  limits?: Record<string, any>;
  price_monthly?: number | null;
  price_annual?: number | null;
  is_custom_price?: boolean;
  stripe_price_id?: string | null;
  stripe_price_id_monthly?: string | null;
  stripe_price_id_annual?: string | null;
  is_active?: boolean;
}) {
  return apiFetch<PlanAdminOut>("console", `/platform/plans/${planId}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  });
}

export async function setTenantPlan(tenantId: string, planKey: string, status: string = "active") {
  return apiFetch<{ ok: boolean; tenant_id: string; plan_key: string; status: string }>(
    "console",
    `/platform/tenants/${tenantId}/plan`,
    {
      method: "PUT",
      body: JSON.stringify({ plan_key: planKey, status }),
    }
  );
}
